lib/SILOptimizer/Utils/ConstantFolding.cpp

```diff
@@ -1761,7 +1763,13 @@ ConstantFolder::processWorkList() {
               FoldedUsers.insert(TI);
           }
 
-          // We were able to fold, so all users should use the new folded value.
+          // We were able to fold, so all users should use the new folded
+          // value. If we don't have any such users, continue.
```

Comment:
I'm not able to see how the comment relates to the code. Does this also break an infinite cycle? If so, what situation would it kick in? I can't envision it.

Reply:
I think DestructureStruct will have a similar issue. That being said, I do need to create a test case for it.

Reply:
Funny story, I started to play with this more. This is the part that actually breaks the cycle. I.e. my current test case just getting rid of this sorts out the issue. I am going to play with the other one a little bit and see if I can get it to loop and also add a better explanation.

Overall though I really dislike the tuple special case here. The problem it is solving will go away once we have multiple result applies and it confuses the logic.

Reply:
I looked into this a bit more. What I found is that the change above triggers the test case that I put in. I think that says that I should try to split this into two different fixes with separate tests.